### DISCUSSION TITLE
if文の条件に一致しなかった場合の処理

### DISCUSSION THREAD

**Discussion by koniooo** [asciidwango/js-primer#1381] [質問]
```js
if (条件式) {
  return foo;
} else {
  return bar;
}
```
上のコードのように```if文```の条件を満たさなかった場合の処理は```else文```に書くのが正しいと思っていたのですが、下のコードのように単に```if文```のブロックの外に書いているコードもよく見かけます。
```js
if (条件式) {
  return foo;
}
return bar;

```
たしかにどちらも同じ結果になるとは思うのですが、パフォーマンスや使い分けの慣習など何か違いがあるのでしょうか？
書籍と直接関係ない、またとても初歩的な質問で申し訳ありません。

**Comment by azu** [comment]
結論的には好みの問題になりますが、コードを読みやすくするための工夫だと思います。

ネストが深いとコードが読みにくく感じるというのは、大体の人が共通している感覚だと思います。
https://jsprimer.net/basic/condition/#nested-if-statement

```js
if (条件式) {
  return foo;
} else {
  return bar;
}

```
もelse部分で不要なブロック(`{ ... }`)があって、これが読みにくいと感じる人もいるのだと思います。

```
else { // これで不要なネストに見える)
  return bar;
}
```

そのため、elseじゃなくて、単純に `return` を外側に書く人たちもいるという感覚です。

```js
if (条件式) {
  return foo;
}
return bar;
```

これを扱うESLintのルールとしてno-else-returnというものがあります。
ただ、このルールも`recommened`ではないので、そういうのを好む人が一定数いるという程度だと思います。

- [no-else-return - Rules - ESLint - Pluggable JavaScript linter](https://eslint.org/docs/rules/no-else-return)

微妙に関連しそうな話としては、Early Return(早期リターン)というテクニックが関係していそうです。

例えば、次のような複雑な条件式を通り抜けたときのみ処理するような場合に、
ifのネストで書くとどこが重要な処理なのかが少しわかりにく感じになります。

```js
// aが条件を全てパスときのみ、実際に処理する
const process(a){ 
  if (条件式) {    
    if (条件式) {
      if (条件式) {
         // 実際に処理する
      } 
    } 
  } 
}
```

これは言い換えればどれかの条件を満たさなかったとき(条件式がfalse)は処理する必要がないということなります。
そこで条件を満たさなかったらすぐにreturnしてしまうことをEarly Returnと言ったりします。
(通常は条件式が複雑なほどコードが複雑になりますが、Early Returnなどのテクニックは条件式が複雑になっても、コードの複雑性は一定に保つ意図があります)

上のコードは、次のように条件式のネストをなくしてフラットにかけます。

```js
// aが条件を全てパスときのみ、実際に処理する
const process(a){ 
  if (!条件式) {
    return;
  } 
  if (!条件式) {
    return;
  } 
  if (!条件式) {
    return;
  } 
  // 実際に処理する
}
```

これもif elseのelseを書かないのと同じようにネストを少し無くそうという書き方として似ているものだと思います。
この話は、JavaScriptに限らず、ifとelseがある大体の言語で同じようなパターンがあるきはします。

そのため、コードをスッキリ見せたいから、余計なブロック(`{}`)を減らして書くためにelseを使っていないという感じだと思います。

次の書籍などで、同じようにネストを減らすテーマがあった気がします。

- [O'Reilly Japan - リーダブルコード](https://www.oreilly.co.jp/books/9784873115658/)
- [リファクタリング 既存のコードを安全に改善する（第2版） | Ohmsha](https://www.ohmsha.co.jp/book/9784274224546/)

**Reply by koniooo** [reply]
非常に丁寧でわかりやすい解説ありがとうございます。
ESLintのルールの一つにもなっていたのですね。
リーダブルコード読んでみます。
